[microdnf] Initial work

jrohel opened the pull request:
The PR is an attempt to implement microdnf based on the new libdnf.
The code is evolving rapidly and can change.

It implements basic skeleton, loads the configuration. It contains basic implementation of the `repolist`, `repoquery`, `install`, `upgrade`, `reinstall`, `remove`, `download` commands.
The PR also implements an argument parser (not all work is done) and more utilities. Some of them will be moved into libraries (libdnf, libdnf-cli).
The PR can contain not merged (work in progress) libdnf code.
The architecture and base architecture detection code may be replaced by an external library (as soon as it exists) in the future.

The PR is too big now. And the code is good enough for testing purposes.

jrohel committed:
[microdnf] - add Base and set MemoryBufferLogger

jrohel committed:
[microdnf] Add utils with am_i_root()

jrohel committed:
[microdnf] Determine log directory and set StreamLogger

jrohel committed:
[microdnf] utils: xdg get user directories

jrohel committed:
[microdnf] Add utils to detect architecture, base arch, and system release

It will probably be reworked acording new requests and moved
to an external library.

Conan-Kudo commented:
@jrohel This will need rebasing now.

jrohel committed:
[microdnf] Add Command base class

The commands will derived from this class.

jrohel committed:
[microdnf] Implmement "repolist" command

jrohel committed:
[microdnf] Basic implementation of "repoquery" command

jrohel committed:
[argument parser] Implementation

Supports:
- long and short named options
- positional options
- command with subcommands
- option conflicts detection
....
In development. Need to implement help message.

May be it will be moved into libdnf-cli.

jrohel committed:
[microdnf] Add Context class

It will collect all the data needed for microdnf.
This commit adds: ArgumentParser, libdnf::Base, container for commands
and selected command.

jrohel committed:
[microdnf] Context: Add support for repository loading

The commit adds Context::load_rpm_repo() method for loading rpm
repository into RepoSack.
It also implements simple repository callback class and userconfirm()
function to confirm key import.

jrohel committed:
[microdnf] Use Context, parse arguments, set and use command

jrohel committed:
[microdnf] Load configurations, detect arch and os release

jrohel committed:
[microdnf] Register "repolist" command

jrohel committed:
[microdnf] Register "repoquery" command

jrohel committed:
[microdnf] Set path to user cache directory

Used when the user effective ID is not 0.

jrohel committed:
[microdnf] Support "--setopt=KEY=VALUE"

jrohel committed:
[microdnf] repoquery: Use rpm::SolvQuery::resolve_pkg_spec()

j-mracek commented on microdnf/commands/repoquery/repoquery.cpp:
```diff
@@ -170,25 +170,21 @@ void CmdRepoquery::run(Context & ctx) {
+    libdnf::rpm::SolvQuery full_solv_query(&solv_sack);
+    for (auto & pattern : *patterns_to_show_options) {
+        libdnf::rpm::SolvQuery solv_query(full_solv_query);
+        solv_query.resolve_pkg_spec(dynamic_cast<libdnf::OptionString *>(pattern.get())->get_value(), true, true, true, true, true, {});
```
Because the performance optimization in dnf-5 is applied only without ICASE, therefore microdnf and dnf should provide a similar performance.

jrohel committed:
[microdnf] Set "cachedir" for root user

jrohel committed:
[microdnf] Add commands install, reinstall, remove, upgrade

Basic implementation. Packages dependencies are not solved.

jrohel committed:
[microdnf] download_packages()

Implemented function for downloading packages.
Commands (install, reinstall, upgrade) now use this function.

jrohel committed:
[microdnf] add command download

Download packages to current directory.

jrohel committed:
[microdnf] PkgDownloadCB class: Use progress bar from libdnf::cli

Used when downloading packages.

jrohel committed:
[microdnf] PkgDownloadCB class: update progress bar state

Updates the status of the progress bar when the package does not need
to be downloaded or an error has occurred. Without this, the indicator
is still held active.

Commit:
[microdnf] Fix output after completing bars

jrohel committed:
[microdnf] Use rpm::Transaction callbacks with progressbars

jrohel committed:
[microdnf] Add Context::load_rpm_repos()

Updates the repositories metadata cache.
Loads the updated metadata into rpm::RepoSack and into rpm::SolvSack.
Uses progress bars during downloading new metadata.
Respects skip_if_unavailable option.

jrohel committed:
[microdnf] add Threads library to linker

jrohel committed:
[microdnf] Context::load_rpm_repos: multithread support

Main thread prepares (updates) repositories metadata. Second thread
loads them to solvable sack.

jrohel committed:
[argument parser] Start help support

jrohel committed:
[microdnf] Use argument parser help support

Conan-Kudo commented on microdnf/commands/repoquery/repoquery.cpp:
```diff
@@ -0,0 +1,188 @@
+    package_info_add_line(table, "Size", std::to_string(size).c_str());
+    package_info_add_line(table, "Source", package.get_sourcerpm().c_str());
+    // TODO(jrohel): support for reponame package_info_add_line(table, "Repository", package.get_reponame().c_str());
+    package_info_add_line(table, "Summanry", package.get_summary().c_str());
```
`Summanry` -> `Summary`

jrohel replied:
Thanks. I fixed the typo.

Conan-Kudo reviewed (changes requested):
One minor spelling error, but otherwise fine, code-wise. However, I wonder why `microdnf/utils` isn't just part of the `libdnf/utils` class?

jrohel commented:
@Conan-Kudo  >However, I wonder why microdnf/utils isn't just part of the libdnf/utils class?

In fact, microdnf is my testing/development platform. I assume that some utilities will be moved into libdnf or libdnf-cli.
The argument parser will also be moved to libdnf-cli.
The question is, move the code right now? Before merging microdnf or later? I'll consider it.

dmach commented:
Let's merge the current code and refactor some pieces to libdnf-cli later.
I don't want to block this from merging any more, it's better to have microdnf available for everyone so they can use it for testing, writing the dnfdaemon-client etc.
